File: algo/maxsum-subarray.c
// Maximal sum sub array.
// GPL(C) moshahmed@gmail.com

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define NN 1000

int main(int argc, char *argv[]){
  int A[NN], N=10;
  int i,j;
  int b, e;
  int sum, max_sum;

  if (argc > 2) { // usage: cmd 13 -3 -25 20 -3 -16 -23 18 20 -7 12 -5 -22 15 -4 7
    N = argc-1;
    for(i=1; i<argc; i++)
      A[i-1] = atoi(argv[i]);
  } else {
    if (argc == 2) { // usage: cmd N
      N = atoi(argv[1]);
    }
    for(i=0;i<N;i++){
      A[i]= 1 + (int ) (N * (float) rand()/(float)RAND_MAX) - N/2 ;
    }
  }

  printf("Array A=[");
  for(i=0;i<N;i++){
    printf("%3d ",A[i]);
    if (i % 12 == 11) printf("\n  ");
  }
  printf("]\n");

  b=-1;
  e=-1;
  max_sum=-INT_MAX;
  for(j=0;j<N;j++){
    sum=0;
    for(i=j;i<N;i++){
      sum+=A[i];
      if(sum>max_sum){
        max_sum=sum;
        b=j;
        e=i;
      }
    }
  }

  printf("max_sum=%d subarray A[%d..%d]=[",max_sum, b, e);
  for(i=b;i<e+1;i++){
    printf("%d ",A[i]);
  }
  printf("]\n");

  return 0;
}

File: algo/maxsum-subarray.R
# What: Maximal sum sub array in R
# 2016-Oct-02 
# Changes GPL(C) moshahmed@gmail.com
# see C:/doc3/algo/malgo/dynamic/msa/maxsum-subarray.c
# $Header: c:/cvs/repo/mosh/stats-r/algo/maxsum-subarray.R,v 1.5 2016-12-04 15:15:23 a Exp $

maximal_sum_subarray <- function(A) {
  N=length(A)
  b=-1
  e=-1
  max_sum=-Inf
  for(j in 0:(N-1)) { # for(j=0;j<N;j++)
    sum=0
    for(i in j:(N-1)) { # for(i=j;i<N;i++)
      sum = sum + A[i+1]
      if(sum>max_sum){
        max_sum=sum
        b=j
        e=i
      }
    }
  }
  cat(sprintf("maximal_sum_subarray of A[%d:%d] is (",b,e),A[(b+1):(e+1)],")")
  cat(" max_sum=",max_sum,"\n")
}

A <- c(13,-3,-25,20,-3,-16,-23,18,20,-7,12,-5,-22,15,-4,7)
cat("A=",A,"\n")
maximal_sum_subarray(A)
